### PR TITLE
PHPBB3-13502 - controller resolver should handle callable functions and objects

### PR DESCRIPTION
**Pull request by rfdy**
This change follows what Symfony does, by allowing the controller to be any type of callable.

### CODE REVIEW

**Comment by nickvergessen** [comment]
I think this was not added intentionally since we want to help extension authors to use DI and therefor have the controllert as classes.

**Comment by rfdy** [comment]
I think I see your point about making extensions easier to write, but I don't see how that is in conflict with DI. In symfony a controller is just any callable that returns a response. That leads to some very useful design patterns. Here is the code I'm using, of course there are other ways to achieve this, but I really don't want the controller to be available publicaly at all, which is why I used an anonymous function.

I use the kernel.controller event to alter the arguments to my method depending on its arguments, similar to how SensioFrameworkExtraBundle ParamConverter works. In my case its for an API I'm building and want to return a very specific response if the argument cannot be instantiated.

```
$event->setController(function() use ($e) {
    return new JsonErrorResponse($e->getMessage());
});
```

Making phpBB work similar to how you'd do things in Symfony seems like a good idea to me as it'll reduce the developer entry barrier. Anyone familiar with Symfony should be able to write phpBB extensions with ease. Changing simple conventions like this could hurt the process.

Thanks.

**Comment by nickvergessen** [comment]
@Nicofuma

**Comment by Nicofuma** [comment]
I see the point....
I'm not sure it's the best design (throwing an exception should do the same thing) but I agree about the Symfony "compliance" so... :+1: for this PR

**Comment by rfdy** [comment]
Throwing an exception would work, but in this case, I really don't want to throw a global exception (ie. it can be thrown anywhere), and then adding another custom exception handler just for this.

Thanks for accepting the patch!

**Comment by Nicofuma** [comment]
Please update the commit message.

By the way maybe you should add that this PR add the support for any callable in the arguments parser but not in the resolver itself. I think it's fine like that for now but maybe we should work on extending the base Resolver in the future (to support the service:method notation and to translate the exceptions)

**Comment by marc1706** [comment]
Please update your commit message to fit the format outlined here: https://wiki.phpbb.com/Git#Commit_Messages

**Comment by Nicofuma** [comment]
bump @rfdy

File: phpBB/phpbb/controller/resolver.php [hunk 1]
```diff
@@ -116,9 +116,18 @@ public function getController(Request $request)
...
 	{
-		// At this point, $controller contains the object and method name
-		list($object, $method) = $controller;
-		$mirror = new \ReflectionMethod($object, $method);
+		// At this point, $controller should be a callable
+		if (is_array($controller))
+		{
+			list($object, $method) = $controller;
+			$mirror = new \ReflectionMethod($object, $method);
+		} else if (is_object($controller) && !$controller instanceof \Closure)
```
Comment (imkingdavid) [inline phpBB/phpbb/controller/resolver.php]: In accordance with the phpBB coding guidelines (read: https://area51.phpbb.com/docs/31x/coding-guidelines.html#codelayout), to keep things consistent, brackets should go on their own line. Please change this one and the two around the `else` on line 128.

**Comment by Nicofuma** [comment]
bump @rfdy

**Pull request closed without merging**